content/courses/program-security/closing-accounts.md

```diff
@@ -1,123 +1,53 @@
+- Implement additional safeguards like force defund instructions to handle edge
+  cases
 
 ## Lesson
 
-While it sounds simple, closing accounts properly can be tricky. There are a
-number of ways an attacker could circumvent having the account closed if you
-don't follow specific steps.
+### Understanding Account Closing and Revival Attacks
 
-To get a better understanding of these attack vectors, let’s explore each of
-these scenarios in depth.
+Closing accounts properly is crucial for maintaining the security and integrity
+of your Solana program. Revival attacks occur when an attacker exploits
+improperly closed accounts to regain control or access unauthorized funds.
 
-### Insecure account closing
+### Real-World Implications: The Wormhole Hack
 
-At its core, closing an account involves transferring its lamports to a separate
-account, thus triggering the Solana runtime to garbage collect the first
-account. This resets the owner from the owning program to the system program.
+The Wormhole bridge hack, which resulted in the loss of 120,000 wETH,
+demonstrated the severe consequences of improper account management. While not
+directly related to account closing, it highlights the importance of robust
+account handling in DeFi protocols.
 
-Take a look at the example below. The instruction requires two accounts:
+### Secure Account Closing Techniques
 
-1. `account_to_close` - the account to be closed
-2. `destination` - the account that should receive the closed account’s lamports
+#### Native Rust Approach
 
-The program logic is intended to close an account by simply increasing the
-`destination` account’s lamports by the amount stored in the `account_to_close`
-and setting the `account_to_close` lamports to 0. With this program, after a
-full transaction is processed, the `account_to_close` will be garbage collected
-by the runtime.
+To securely close an account:
 
-```rust
-use anchor_lang::prelude::*;
-
-declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
+1. Transfer the account's lamports to a designated receiver
+2. Zero out the account's data
+3. Set a closed account discriminator
 
-#[program]
-pub mod closing_accounts_insecure {
-    use super::*;
-
-    pub fn close(ctx: Context<Close>) -> ProgramResult {
-        let dest_starting_lamports = ctx.accounts.destination.lamports();
-
-        **ctx.accounts.destination.lamports.borrow_mut() = dest_starting_lamports
-            .checked_add(ctx.accounts.account_to_close.to_account_info().lamports())
-            .unwrap();
-        **ctx.accounts.account_to_close.to_account_info().lamports.borrow_mut() = 0;
-
-        Ok(())
-    }
-}
-
-#[derive(Accounts)]
-pub struct Close<'info> {
-    account_to_close: Account<'info, Data>,
-    destination: AccountInfo<'info>,
-}
-
-#[account]
-pub struct Data {
-    data: u64,
-}
 ```
-
-However, the garbage collection doesn't occur until the transaction completes.
```

Comment:
Can you explain more about why you removed the text about how this vulnerability could be exploited?